compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -425,71 +456,77 @@ MMAScheduleAttr::getContractionLayout(vector::ContractionOp contractOp) const {
   SmallVector<int64_t, 2> bPermute = {bK, bN};
   SmallVector<int64_t, 2> cPermute = {cM, cN};
 
-  // TODO: drop this and permute the following fields.
-  if (!isIdentityPermutation(aPermute) || !isIdentityPermutation(bPermute) ||
-      !isIdentityPermutation(cPermute))
-    return std::nullopt;
-
   auto mfmaAttr = llvm::cast<MFMAAttr>(getIntrinsic());
+  MLIRContext *context = getContext();
 
-  // TODO: revisit the handling of subgroup/thread basis.
-  SmallVector<int64_t, 2> subgroupBasis = {getSubgroupMCount(),
-                                           getSubgroupNCount()};
+  // Get the concrete nested layout for each matrix. Note that the struct
+  // MFMAAttr::SignalSubgroupLayout contains the partial layout for the
+  // canonical (M, K) x (K, N) -> (M, N) matmul form; while the specific
+  // contract op we are looking at right now may not be exactly in that form.
+  // So here we need to transpose/permuate the canonical layout to match with
+  // the concrete contract op.
 
   // C matrix layout
   MFMAAttr::SingleSubgroupLayout cCounts =
       mfmaAttr.getCSingleSubgroupLayoutCount();
   MFMAAttr::SingleSubgroupLayout cOrders =
       mfmaAttr.getCSingleSubgroupLayoutOrder();
 
-  SmallVector<int64_t, 2> cSubgroupPerWorkgroup = {getSubgroupMCount(),
-                                                   getSubgroupNCount()};
-  SmallVector<int64_t, 2> cBatchesPerSubgroup = {getSubgroupMTileCount(),
-                                                 getSubgroupNTileCount()};
-  SmallVector<int64_t, 2> cSubgroupOrder = {0, 1};
-  SmallVector<int64_t, 2> cBatchOrder = {0, 1};
-  SmallVector<int64_t, 2> cThreadBasis = cCounts.thread;
-
-  auto cLayout = NestedLayoutAttr::get(
-      getContext(), cSubgroupPerWorkgroup, cSubgroupOrder, cBatchesPerSubgroup,
-      cBatchOrder, cCounts.outer, cOrders.outer, cCounts.thread, cOrders.thread,
-      cCounts.element, cOrders.element, subgroupBasis, cThreadBasis);
+  // Note that no matter how we permute/transpose the input contraction problem,
+  // the way we view the hardware warps/threads remain the same--that is, from
+  // hardware's perspective, a single thread has the same warp and thread ID no
+  // matter what part of the contraction data it works on. Similarly here, we
+  // are delinearizing the linearized GPU hardware thread ID into a n-D
+  // concatenated logical warp+thread using the basis, and they should remain
+  // the same for all A/B/C matrix.
+  SmallVector<int64_t, 2> subgroupBasis = {getSubgroupMCount(),
+                                           getSubgroupNCount()};
+  SmallVector<int64_t, 2> threadBasis = cCounts.thread;
```

Comment:
I think this needs to be the following
```suggestion
  SmallVector<int64_t, 2> threadBasis = applyPermutation(cCounts.thread, cOrders.thread);
```
Because the `thread_order` field indicates which thread counts are outermost to innermost. The basis needs to be the sizes to delinearize to independent of the thread order. For example,

```
thread_order = [2, 0, 1]
thread_sizes = [5, 3, 7]
```

This means that the outermost thread id has range `[0, 7)`, the middle thread id has range `[0, 5)` and the innermost thread id has range `[0, 3)`. So to get that basis (`[7, 5, 3]`) from the thread sizes we need to apply the permutation. I think we are most likely still mixing up `order` and `inverse(order)` in some places though and are getting lucky because all of our examples are strictly 2D.

Reply:
Good catch! And thanks for walking me through it. I think this makes sense! Fixed.